unblob/handlers/archive/cpio.py

```diff
@@ -253,19 +344,128 @@ def _calculate_name_size(header) -> int:
     def _calculate_mode(header) -> int:
         return decode_int(header.c_mode, 16)
 
+    @staticmethod
+    def _calculate_dev(header) -> int:
+        return os.makedev(
+            decode_int(header.c_dev_maj, 16), decode_int(header.c_dev_min, 16)
+        )
+
+    @staticmethod
+    def _calculate_rdev(header) -> int:
+        return os.makedev(
+            decode_int(header.c_rdev_maj, 16), decode_int(header.c_rdev_min, 16)
+        )
+
+
+class PortableASCIIWithCRCParser(PortableASCIIParser):
+    def valid_checksum(self, header, start_offset: int) -> bool:
+        header_checksum = decode_int(header.c_chksum, 16)
+        calculated_checksum = 0
+        file_size = self._calculate_file_size(header)
 
-class PortableASCIIWithCRCHandler(_NewASCIICommon, _CPIOHandlerBase):
+        for chunk in iterate_file(self.file, start_offset, file_size):
+            calculated_checksum += sum(bytearray(chunk))
+
+        return header_checksum == calculated_checksum
+
+
+class _CPIOExtractorBase(Extractor):
+    PARSER: Type[CPIOParserBase]
+
+    def extract(self, inpath: Path, outdir: Path) -> Optional[ExtractResult]:
+        fs = FileSystem(outdir)
+
+        with File.from_path(inpath) as file:
+            parser = self.PARSER(file, 0)
+            parser.parse()
+            parser.dump_entries(fs)
+
+
+class BinaryCPIOExtractor(_CPIOExtractorBase):
+    PARSER = BinaryCPIOParser
+
+
+class PortableOldASCIIExtractor(_CPIOExtractorBase):
+    PARSER = PortableOldASCIIParser
+
+
+class PortableASCIIExtractor(_CPIOExtractorBase):
+    PARSER = PortableASCIIParser
+
+
+class PortableASCIIWithCRCExtractor(_CPIOExtractorBase):
+    PARSER = PortableASCIIWithCRCParser
+
+
+class _CPIOHandlerBase(Handler):
+    """A common base for all CPIO formats.
+
+    The format should be parsed the same, there are small differences how to calculate
+    file and filename sizes padding and conversion from octal / hex.
+    """
+
+    EXTRACTOR = None
+
+    def calculate_chunk(self, file: File, start_offset: int) -> Optional[ValidChunk]:
+        raise NotImplementedError
```

Comment:
I think the current code is clean, there is only the redundant looking implementation of `calculate_chunk()` for every CPIO `Handler` below.

Maybe they could be eliminated by implementing a generic one here:

```suggestion
    EXTRACTOR: _CPIOExtractorBase

    def calculate_chunk(self, file: File, start_offset: int) -> Optional[ValidChunk]:
        parser = self.EXTRACTOR.PARSER(file, start_offset)
        parser.parse()
        return ValidChunk(
            start_offset=start_offset,
            end_offset=parser.end_offset,
        )
```